File: packages/provide-uterm-go/screen/pytext.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package screen

import (
	"strings"
	"unicode"
)

// isPySpace reports whether r matches Python's re \s (equivalently
// str.isspace) for str patterns. That set is Go's unicode.IsSpace plus the
// information separators U+001C-U+001F (bidirectional classes B/S), which
// lack the Unicode White_Space property and are therefore missed by
// unicode.IsSpace.
func isPySpace(r rune) bool {
	return unicode.IsSpace(r) || (r >= 0x1c && r <= 0x1f)
}

// isPyDigit reports whether r matches Python's re \d for str patterns:
// Unicode decimal digits (category Nd), which is exactly Go's
// unicode.IsDigit.
func isPyDigit(r rune) bool {
	return unicode.IsDigit(r)
}

// pyStrip mirrors Python's str.strip() with no arguments (trims the
// str.isspace character set from both ends).
func pyStrip(s string) string {
	return strings.TrimFunc(s, isPySpace)
}

// isPyLineTerminator reports whether r is one of the line boundaries
// recognized by Python's str.splitlines().
func isPyLineTerminator(r rune) bool {
	switch r {
	case '\n', '\r', '\v', '\f', 0x1c, 0x1d, 0x1e, 0x85, 0x2028, 0x2029:
		return true
	}
	return false
}

// pySplitLines mirrors Python's str.splitlines(): it splits on the full set
// of Python line boundaries (treating "\r\n" as a single boundary), drops
// the terminators, and yields no trailing empty line. An empty string
// yields an empty slice.
func pySplitLines(s string) []string {
	rs := []rune(s)
	lines := []string{}
	start := 0
	for i := 0; i < len(rs); {
		if !isPyLineTerminator(rs[i]) {
			i++
			continue
		}
		lines = append(lines, string(rs[start:i]))
		if rs[i] == '\r' && i+1 < len(rs) && rs[i+1] == '\n' {
			i++
		}
		i++
		start = i
	}
	if start < len(rs) {
		lines = append(lines, string(rs[start:]))
	}
	return lines
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/Screen/PyText.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Text;

namespace Provide.Uterm.Screen;

internal static class PyText
{
    internal static bool IsPySpace(int r) =>
        char.IsWhiteSpace((char)r) && r <= 0xFFFF
            ? RuneWhiteSpace(r)
            : RuneWhiteSpace(r);

    private static bool RuneWhiteSpace(int r)
    {
        if (r >= 0 && r <= char.MaxValue && char.IsWhiteSpace((char)r))
        {
            return true;
        }

        // Unicode White_Space property for non-BMP is rare; also include 0x1c-0x1f
        if (r >= 0x1c && r <= 0x1f)
        {
            return true;
        }

        // Use Rune.IsWhiteSpace for general case
        try
        {
            return Rune.IsWhiteSpace(new Rune(r));
        }
        catch
        {
            return false;
        }
    }

    internal static bool IsPyDigit(int r) =>
        r <= char.MaxValue ? char.IsDigit((char)r) : Rune.IsDigit(new Rune(r));

    internal static string PyStrip(string s)
    {
        var runes = s.EnumerateRunes().ToList();
        var start = 0;
        var end = runes.Count;
        while (start < end && IsPySpace(runes[start].Value))
        {
            start++;
        }

        while (end > start && IsPySpace(runes[end - 1].Value))
        {
            end--;
        }

        if (start == 0 && end == runes.Count)
        {
            return s;
        }

        return string.Concat(runes.Skip(start).Take(end - start).Select(r => r.ToString()));
    }

    private static bool IsPyLineTerminator(int r) =>
        r is '\n' or '\r' or '\v' or '\f' or 0x1c or 0x1d or 0x1e or 0x85 or 0x2028 or 0x2029;

    internal static List<string> PySplitLines(string s)
    {
        var rs = s.EnumerateRunes().Select(r => r.Value).ToList();
        var lines = new List<string>();
        var start = 0;
        for (var i = 0; i < rs.Count;)
        {
            if (!IsPyLineTerminator(rs[i]))
            {
                i++;
                continue;
            }

            lines.Add(string.Concat(rs.Skip(start).Take(i - start).Select(r => char.ConvertFromUtf32(r))));
            if (rs[i] == '\r' && i + 1 < rs.Count && rs[i + 1] == '\n')
            {
                i++;
            }

            i++;
            start = i;
        }

        if (start < rs.Count)
        {
            lines.Add(string.Concat(rs.Skip(start).Select(r => char.ConvertFromUtf32(r))));
        }

        return lines;
    }
}
